### ISSUE TITLE
php bakery migrate:rollback missing class errors

### ISSUE THREAD

**Issue by amosfolz** [userfrosting/UserFrosting#1002]
Trying to run `php bakery migrate:rollback` will fail if there is a missing migration class, even if it is not in the last batch that was ran. 

![image](https://user-images.githubusercontent.com/33728190/60052886-7730ca80-96a4-11e9-8f3b-23a088720fca.png)

![image](https://user-images.githubusercontent.com/33728190/60052900-8283f600-96a4-11e9-8527-0b9ebd4d3a75.png)

It seems it shouldn't matter if a migration class is missing unless that class is part of the last batch ran, since `php bakery migrate:rollback` just reverses the most recent migration. Unless `migrations` != `batches`?

**Comment by lcharette** [comment]
Can you pinpoint where this exception is thrown ?

**Comment by amosfolz** [comment]
> 
> 
> Can you pinpoint where this exception is thrown ?

It seems to be from this [line](https://github.com/userfrosting/UserFrosting/blob/master/app/sprinkles/core/src/Database/Migrator/MigrationDependencyAnalyser.php#L213)

Rather than trying to modify anything in any of the `Migrator` classes, a simpler solution might be to ask the user if they want to drop the "stale" record from the database table.

**Comment by lcharette** [comment]
That might be a good idea... `migrate:clean` ?

**Comment by amosfolz** [comment]
```
#0  UserFrosting\Sprinkle\Core\Database\Migrator\MigrationDependencyAnalyser->getMigrationDependencies() called at [/home/vagrant/userfrosting/app/sprinkles/core/src/Database/Migrator/MigrationRollbackDependencyAnalyser.php:66]
#1  UserFrosting\Sprinkle\Core\Database\Migrator\MigrationRollbackDependencyAnalyser->validateClassDependencies() called at [/home/vagrant/userfrosting/app/sprinkles/core/src/Database/Migrator/MigrationDependencyAnalyser.php:78]
#2  UserFrosting\Sprinkle\Core\Database\Migrator\MigrationDependencyAnalyser->analyse() called at [/home/vagrant/userfrosting/app/sprinkles/core/src/Database/Migrator/MigrationDependencyAnalyser.php:161]
#3  UserFrosting\Sprinkle\Core\Database\Migrator\MigrationDependencyAnalyser->getUnfulfillable() called at [/home/vagrant/userfrosting/app/sprinkles/core/src/Database/Migrator/Migrator.php:327]
#4  UserFrosting\Sprinkle\Core\Database\Migrator\Migrator->checkRollbackDependencies() called at [/home/vagrant/userfrosting/app/sprinkles/core/src/Database/Migrator/Migrator.php:283]
#5  UserFrosting\Sprinkle\Core\Database\Migrator\Migrator->rollbackMigrations() called at [/home/vagrant/userfrosting/app/sprinkles/core/src/Database/Migrator/Migrator.php:218]
#6  UserFrosting\Sprinkle\Core\Database\Migrator\Migrator->rollback() called at [/home/vagrant/userfrosting/app/sprinkles/core/src/Bakery/MigrateRollbackCommand.php:80]
#7  UserFrosting\Sprinkle\Core\Bakery\MigrateRollbackCommand->execute() called at [/home/vagrant/userfrosting/app/vendor/symfony/console/Command/Command.php:255]
#8  Symfony\Component\Console\Command\Command->run() called at [/home/vagrant/userfrosting/app/vendor/symfony/console/Application.php:921]
#9  Symfony\Component\Console\Application->doRunCommand() called at [/home/vagrant/userfrosting/app/vendor/symfony/console/Application.php:273]
#10 Symfony\Component\Console\Application->doRun() called at [/home/vagrant/userfrosting/app/vendor/symfony/console/Application.php:149]
#11 Symfony\Component\Console\Application->run() called at [/home/vagrant/userfrosting/app/system/Bakery/Bakery.php:82]
#12 UserFrosting\System\Bakery\Bakery->run() called at [/home/vagrant/userfrosting/bakery:35]
```

**Comment by amosfolz** [comment]
Okay, based off the comment I found [here](https://github.com/userfrosting/UserFrosting/blame/master/app/sprinkles/core/src/Database/Migrator/Migrator.php#L292-L298), it seems the intention is to actually have the rollback continue in these cases of missing classes...

**Comment by lcharette** [comment]
I remember that comment... At some point there was no way to display a warning from the migrator to the CLI as the migrator could be run independently from the CLI. I _think_ the note thing solved that issue? 

As long as the warning is displayed, I guess it should be fine to skip the migration. It might still break stuff, but nothing more can be done anyway if the class doesn't exist anymore...

**Comment by amosfolz** [comment]
There might be other situations where that note is still needed but when running `migrate:rollback` missing  classes will cause [`Migrator->checkRollbackDependencies()`](Migrator->checkRollbackDependencies()) to throw an exception.  


On another note, I was looking through the debug trace I posted above and I am a bit confused about something. At line #2:
```
#2  UserFrosting\Sprinkle\Core\Database\Migrator\MigrationDependencyAnalyser->analyse()
```
Inside `analyse()` method there is this line [`$this->validateClassDependencies($migration);`](http://api.userfrosting.com/source-class-UserFrosting.Sprinkle.Core.Database.Migrator.MigrationDependencyAnalyser.html#78)

I would have expected this to call the method at [this line](http://api.userfrosting.com/source-class-UserFrosting.Sprinkle.Core.Database.Migrator.MigrationDependencyAnalyser.html#97) inside the `MigrationDependencyAnalyser` class. 

But if you look at the stack trace:
```
#1  UserFrosting\Sprinkle\Core\Database\Migrator\MigrationRollbackDependencyAnalyser->validateClassDependencies() 
```
It actually calls the same method but from the [`MigrationRollbackDependencyAnalyser`](http://api.userfrosting.com/source-class-UserFrosting.Sprinkle.Core.Database.Migrator.MigrationRollbackDependencyAnalyser.html#46)

I am just curious as to why that happens here.

**Comment by lcharette** [comment]
Fixed in 4.2.3 by PR #1007
